lib/tree.rb

```diff
@@ -16,45 +16,127 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn) if the tree is balanced (best case), O(n) if tree is unbalanced (worst case)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    if !@root
+      new_node = TreeNode.new(key, value)
+      @root = new_node
+      return
+    end
+
+    current = @root
+    new_node = TreeNode.new(key, value)
+
+    until !current
+      if new_node.key <= current.key # GO LEFT
+        if !current.left
+          current.left = new_node
+          return
+        else
+          current = current.left
+        end
+      else # new_node.value > current.value # GO RIGHT
+        if !current.right
+          current.right = new_node
+          return
+        else
+        current = current.right
+        end
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn) where n is the length of the array if the tree is balanced. O(n) if not.
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return nil if !@root
+    current = @root
+
+    until current == nil
+      return current.value if key == current.key
+      key < current.key ? current = current.left : current = current.right
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n-squared) 
+  # Space Complexity: O(n) where n is the height of the recursion stack
   def inorder
-    raise NotImplementedError
+    @output = []
+    current = @root
+
+    return @output if !@root
+
+    def analyzeInorder(current)
```

Comment:
An inner method?  I suggest either putting this method in `TreeNode` or making it a private helper method. Output doesn't need to be an instance variable.